### PR TITLE
a proof of concept implementation for shared start txn calls

### PR DESCRIPTION
**Pull request by gokcanozakdag**
**Goals (and why)**:
To demonstrate one of the ways of implementing shared start txn calls.

**Implementation Description (bullets)**:

**Testing (What was existing testing like?  What have you done to improve it?)**:

**Concerns (what feedback would you like?)**:

**Where should we start reviewing?**:

**Priority (whenever / two weeks / yesterday)**:

### CODE REVIEW

File: lock-api/src/main/java/com/palantir/lock/client/TimelockTransactionService.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import com.palantir.lock.v2.StartIdentifiedAtlasDbTransactionResponse;
+
+//better name
+
+/**
+ * responsible for managing shared start txn calls
+ */
+public interface TimelockTransactionService {
+    StartIdentifiedAtlasDbTransactionResponse startTransaction(); //should coalesce concurrent calls
+    Set<LockToken> reduceForRefresh(Set<LockToken> tokens);
```
Comment (gokcanozakdag) [inline lock-api/src/main/java/com/palantir/lock/client/TimelockTransactionService.java]: these are just static methods, probably should be moved to somewhere else

File: lock-api/src/main/java/com/palantir/lock/client/SimpleTimelockTransactionService.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+public class SimpleTimelockTransactionService implements TimelockTransactionService {
+    private LeasingTimelockClient delegate;
+
+    public SimpleTimelockTransactionService(LeasingTimelockClient delegate) {
+        this.delegate = delegate;
+    }
+
+
+    @Override
+    public StartIdentifiedAtlasDbTransactionResponse startTransaction() {
```
Comment (gokcanozakdag) [inline lock-api/src/main/java/com/palantir/lock/client/SimpleTimelockTransactionService.java]: simpleTxnService just delegates the startTxn call (and wraps it) without coalescing. CoalescingTxnService will be using AutoBatcher to batch these calls into a single startTxns call (this endpoint doesn't exist right now)

File: lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockClient.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        return leasingTimelockClient.lock(request);
+    }
+
+    @Override
+    public WaitForLocksResponse waitForLocks(WaitForLocksRequest request) {
+        return rawClient.waitForLocks(request);
+    }
+
+    @Override
+    public Set<LockToken> refreshLockLeases(Set<LockToken> tokens) {
```
Comment (j-baker) [inline lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockClient.java]: i think these two methods are both critically broken

File: lock-api/src/main/java/com/palantir/lock/client/SharedImmutableTsToken.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        //precondition
+        reference--;
+    }
+
+    synchronized boolean dereferenced() {
+        return reference == 0;
+    }
+
+    LockToken token() {
+        return token;
```
Comment (j-baker) [inline lock-api/src/main/java/com/palantir/lock/client/SharedImmutableTsToken.java]: this should probably throw if the lock is already reference counted out

File: lock-api/src/main/java/com/palantir/lock/client/SharedImmutableTsToken.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+import java.util.UUID;
+
+import com.palantir.lock.v2.LockToken;
+
+class SharedImmutableTsToken implements LockToken {
+    private LockToken token;
+    private volatile int reference;
+
+    SharedImmutableTsToken(LockToken token) {
```
Comment (j-baker) [inline lock-api/src/main/java/com/palantir/lock/client/SharedImmutableTsToken.java]: think it's worth this taking the reference count in its constructor and only allowing it to be decremented. It means you can't get to zero and then go back up to 1

File: lock-api/src/main/java/com/palantir/lock/client/SharedImmutableTsToken.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    synchronized void mark() {
+        reference++;
+    }
+
+    synchronized void unmark() {
+        //precondition
+        reference--;
+    }
+
+    synchronized boolean dereferenced() {
```
Comment (j-baker) [inline lock-api/src/main/java/com/palantir/lock/client/SharedImmutableTsToken.java]: this returns the wrong result if you call unmark too many times

File: lock-api/src/main/java/com/palantir/lock/client/SharedImmutableTsToken.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    SharedImmutableTsToken(LockToken token) {
+        this.token = token;
+        reference = 0;
+    }
+
+    synchronized void mark() {
+        reference++;
+    }
+
+    synchronized void unmark() {
```
Comment (j-baker) [inline lock-api/src/main/java/com/palantir/lock/client/SharedImmutableTsToken.java]: probably call this release

File: lock-api/src/main/java/com/palantir/lock/client/SharedImmutableTsToken.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+package com.palantir.lock.client;
+
+import java.util.UUID;
+
+import com.palantir.lock.v2.LockToken;
+
+class SharedImmutableTsToken implements LockToken {
+    private LockToken token;
+    private volatile int reference;
```
Comment (j-baker) [inline lock-api/src/main/java/com/palantir/lock/client/SharedImmutableTsToken.java]: this doesn't need to be volatile

File: lock-api/src/main/java/com/palantir/lock/client/SimpleTimelockTransactionService.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    @Override
+    public Set<LockToken> reduceForUnlock(Set<LockToken> tokens) {
+        Set<HeldImmutableTsToken> immutableTsTokens = filterImmutableTsTokens(tokens);
+        immutableTsTokens.forEach(HeldImmutableTsToken::invalidate);
+        Set<LockToken> toUnlock;
+        synchronized (this) {
+            toUnlock = immutableTsTokens.stream().map(h -> h.sharedToken())
+                    .filter(SharedImmutableTsToken::dereferenced)
```
Comment (j-baker) [inline lock-api/src/main/java/com/palantir/lock/client/SimpleTimelockTransactionService.java]: don't think there's any place these ever get dereferenced?

File: lock-api/src/main/java/com/palantir/lock/client/SimpleTimelockTransactionService.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    private Set<HeldImmutableTsToken> filterImmutableTsTokens(Set<LockToken> tokens) {
+        return tokens.stream().filter(t -> t instanceof HeldImmutableTsToken)
+                .map(t -> (HeldImmutableTsToken) t)
+                .collect(Collectors.toSet());
+    }
+
+    private Set<LockToken> filterOutImmutableTsTokens(Set<LockToken> tokens) {
+        return tokens.stream().filter(t -> !(t instanceof HeldImmutableTsToken))
```
Comment (j-baker) [inline lock-api/src/main/java/com/palantir/lock/client/SimpleTimelockTransactionService.java]: probably worth having a function boolean isImmutableLockToken?

File: lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockClient.java (hunk 3 above)
Comment (j-baker) [reply lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockClient.java]: worth thinking through what the contract of these methods is

File: lock-api/src/main/java/com/palantir/lock/client/HeldImmutableTsToken.java [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+package com.palantir.lock.client;
+
+import java.util.UUID;
+
+import com.palantir.lock.v2.LockToken;
+
+class HeldImmutableTsToken implements LockToken {
+    private UUID requestId;
+    private SharedImmutableTsToken sharedImmutableTsToken;
+    private volatile boolean invalidated;
```
Comment (j-baker) [inline lock-api/src/main/java/com/palantir/lock/client/HeldImmutableTsToken.java]: probably doesn't need to be invalidated

File: lock-api/src/main/java/com/palantir/lock/client/SimpleTimelockTransactionService.java (hunk 9 above)
Comment (j-baker) [reply lock-api/src/main/java/com/palantir/lock/client/SimpleTimelockTransactionService.java]: ah, i found it

File: lock-api/src/main/java/com/palantir/lock/client/SimpleTimelockTransactionService.java [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+        Set<HeldImmutableTsToken> immutableTsTokens = filterImmutableTsTokens(tokens);
+        Set<LockToken> reducedImmutableTsTokens = immutableTsTokens.stream()
+                .map(h -> h.sharedToken().token())
+                .collect(Collectors.toSet());
+
+        return Sets.union(filterOutImmutableTsTokens(tokens), reducedImmutableTsTokens);
+    }
+
+    @Override
+    public Set<LockToken> reduceForUnlock(Set<LockToken> tokens) {
```
Comment (j-baker) [inline lock-api/src/main/java/com/palantir/lock/client/SimpleTimelockTransactionService.java]: This code is very... I don't know the right word. To me, it does a bit too much reaching inside the state of other code in order to figure out what to do; it means that code that should be owned somewhere else is owned here.

As an example, why doesn't unlock on HeldImmutableTsToken return an `Optional<LockToken>` which is the LockToken which should be passed to the underlying lock service? and it does that only the last time? In that case this class doesn't have to be aware of the shared tokens, and the synchronized can vanish :)

File: lock-api/src/main/java/com/palantir/lock/client/SimpleTimelockTransactionService.java [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+        Set<HeldImmutableTsToken> immutableTsTokens = filterImmutableTsTokens(tokens);
+        immutableTsTokens.forEach(HeldImmutableTsToken::invalidate);
+        Set<LockToken> toUnlock;
+        synchronized (this) {
```
Comment (j-baker) [inline lock-api/src/main/java/com/palantir/lock/client/SimpleTimelockTransactionService.java]: red flag for you: a synchronized call which doesn't wrap calls to multiple fields probably implies you're synchronizing in the wrong place.

**Pull request closed without merging**